src/main/kotlin/com/github/treepat/match_manager/MatchManager.kt

```diff
@@ -0,0 +1,64 @@
+package com.github.treepat.match_manager
+
+import com.github.treepat.expression.TreepatExpression
+import com.github.treepat.target_tree.TargetTree
+import com.github.treepat.target_tree.TargetTreeNode
+
+class MatchManager(private var expression: TreepatExpression, private var tree: TargetTree) {
+
+    private var nextNodeToEvaluate: TargetTreeNode? = null
+
+    init {
+        this.nextNodeToEvaluate = tree.root
+    }
+
+    fun getNextMatch(): List<TargetTreeNode>? {
+        if (nextNodeToEvaluate == null) {
+            nextNodeToEvaluate = tree.root
+            return null
+        }
+
+        var response = expression.executeExpression(nextNodeToEvaluate)
+        this.nextNodeToEvaluate = nextNodeToEvaluate!!.nextPreorderNode()
+
+        while ((!response.hasMatch || response.responses.filter { it.matches.isNotEmpty() }.isEmpty()) && nextNodeToEvaluate != null) {
```

Comment:
```suggestion
response.responses.filter { it.matches.isNotEmpty() }.isEmpty())
```
Please use a variable for this, you call this filter two times.

Reply:
I agree, it wont change the functionality but it would defiantly make the code more readable. :)

Reply:
Fair enough